### DISCUSSION TITLE
Collect planned breaking changes for the next major release

### DISCUSSION THREAD

**Discussion by chenjiahan** [web-infra-dev/modern.js#4119] [General]
Although Modern.js has no plans to release v3.0 in 2023, we want to collect planned breaking changes for the next major release during the daily iteration. And these changes should benefit the architectural evolution and long-term maintenance of Modern.js.

This thread will be used to record and discuss all planned breaking changes, so that when we do release v3.0, those breaking changes will be applied to it.

**Comment by chenjiahan** [comment]
Remove compatibility code for Modern.js v1, such as `defineLegacyConfig`.

**Comment by chenjiahan** [comment]
Deprecate storybook v6.

And storybook v7 should be supported in Modern.js 2.x

**Comment by chenjiahan** [comment]
Remove the deprecated exports of Builder plugins, such as https://github.com/web-infra-dev/modern.js/blob/main/packages/builder/plugin-node-polyfill/src/index.ts#L87-L90

**Comment by h-a-n-a** [comment]
Rspack team would like to drop the support for "transpiling everything" in this quater.

I believe there will be some projects creating with Modern.js and `Rspack` and these 
projects might be relying on some third-party packages. For these projects, I believe
they do not configure the `source.include` to manually transpile some of the dependencies.

As soon as we drop the support, the problem that mentioned above would be a breaking 
change to all those projects. I haven't come up with a good idea to notify every users.
Any thoughts? @chenjiahan @web-infra-dev/rspack-core

Related: https://github.com/web-infra-dev/rspack/issues/3067

**Reply by 9aoy** [reply]
yeah, this is unaligned behavior with Modern.js webpack mode. We should not apply this behavior default anymore.
I prefer to add a new configuration to keep "transpiling everything", just like "experimental-rspack-legacy". And in the next major version of Modern.js, we can remove all legacy supports.

```diff
export default defineConfig<'rspack'>({
  plugins: [
    appTools({
-       bundler: 'experimental-rspack'
+      bundler: 'experimental-rspack-legacy',
    }),
  ],
});

**Reply by h-a-n-a** [reply]
@9aoy This sounds fascinating! I love this solution. 

On top of that, we might have also provide a tool for diagnosing current output artifacts to help users to migrate from legacy version to the latest version smoothly. Like a language feature detection tool to make sure we are not breaking anything related to legacy clients.

**Reply by hardfist** [reply]
there're more breaking change other than this, we can't  always change the bundler name

**Reply by 9aoy** [reply]
@h-a-n-a Modern.js supports syntax compatibility checking for output artifacts through the configuration of ﻿security.checkSyntax. However, this feature is disabled by default.
https://modernjs.dev/builder/en/api/config-security.html#securitychecksyntax

**Reply by 9aoy** [reply]
@hardfist Maybe we can add a separate configuration to configure the behavior that needs to be compatible or requires a compatible version of rspack. like ` keepRspackLegacies: ['transpileAll', 'xxxx'],` or ` inheritRspackDefalutBehavior: '0.2',`

**Comment by chenjiahan** [comment]
Upgrade to styled-components v6:

- styled-components v6: https://github.com/styled-components/styled-components/releases/tag/v6.0.0
- babel-plugin-styled-components v2: https://github.com/styled-components/babel-plugin-styled-components/releases/tag/v2.0.0

**Comment by chenjiahan** [comment]
Dropping support for both Node 14 and Node 16 since they are EOL

**Comment by chenjiahan** [comment]
Remove `source.designSystem`, see https://github.com/web-infra-dev/modern.js/discussions/4509 for more details

**Comment by chenjiahan** [comment]
Consider compile all TypeScript files by default, keep JS and TS compile rules separate.
